### PR TITLE
Merge sort docs and casts

### PR DESCRIPTION
**Pull request by gevtushenko**
This PR contains:
1) Documentation refactoring
2) Types clarifications
3) Direct usage of `CUB_NS_QUALIFIER` in the merge sort dispatch mechanism

### CODE REVIEW

File: cub/device/device_merge_sort.cuh [hunk 1]
```diff
@@ -126,45 +133,91 @@ struct DeviceMergeSort
...
+   *
+   * @tparam KeyIteratorT
+   *   is a model of [Random Access Iterator], `KeyIteratorT` is mutable, and
+   *   `KeyIteratorT`'s `value_type` is a model of [LessThan Comparable], and
+   *   the ordering relation on `KeyIteratorT`'s `value_type` is a
+   *   *strict weak ordering*, as defined in the [LessThan Comparable]
+   *   requirements
+   *
+   * @tparam ValueIteratorT
+   *   is a model of [Random Access Iterator], `ValueIteratorT` is mutable, and
```
Comment (zasdfgbnm) [inline cub/device/device_merge_sort.cuh:161]: I think values don't need to be comparable?

File: cub/device/device_merge_sort.cuh [hunk 2]
```diff
@@ -211,61 +267,117 @@ struct DeviceMergeSort
...
+   *   [LessThan Comparable], and the ordering relation on `KeyInputIteratorT`'s
+   *   `value_type` is a *strict weak ordering*, as defined in the
+   *   [LessThan Comparable] requirements
+   *
+   * @tparam ValueInputIteratorT
+   *   is a model of [Random Access Iterator], `ValueInputIteratorT` is mutable,
+   *   and `ValueInputIteratorT`'s `value_type` is a model of
+   *   [LessThan Comparable], and the ordering relation on
+   *   `ValueInputIteratorT`'s `value_type` is a *strict weak ordering*, as
+   *   defined in the [LessThan Comparable] requirements
```
Comment (zasdfgbnm) [inline cub/device/device_merge_sort.cuh:301]: Same here, values don't need to be comparable?

File: cub/device/device_merge_sort.cuh [hunk 3]
```diff
@@ -211,61 +267,117 @@ struct DeviceMergeSort
...
+   *   the ordering relation on `KeyIteratorT`'s `value_type` is a
+   *   *strict weak ordering*, as defined in the [LessThan Comparable]
+   *   requirements
+   *
+   * @tparam ValueIteratorT
+   *   is a model of [Random Access Iterator], `ValueIteratorT` is mutable, and
+   *   `ValueIteratorT`'s `value_type` is a model of [LessThan Comparable], and
+   *   the ordering relation on `ValueIteratorT`'s `value_type` is a
+   *   *strict weak ordering*, as defined in the [LessThan Comparable]
+   *   requirements
```
Comment (zasdfgbnm) [inline cub/device/device_merge_sort.cuh:315]: Same here, values don't need to be comparable?

File: cub/device/device_merge_sort.cuh [hunk 4]
```diff
@@ -474,46 +679,90 @@ struct DeviceMergeSort
...
+   *   and the ordering relation on `KeyIteratorT`'s `value_type` is a
+   *   *strict weak ordering*, as defined in the [LessThan Comparable]
+   *   requirements
+   *
+   * @tparam ValueIteratorT
+   *   is a model of [Random Access Iterator], `ValueIteratorT` is mutable, and
+   *   `ValueIteratorT`'s `value_type` is a model of [LessThan Comparable],
+   *   and the ordering relation on `ValueIteratorT`'s `value_type` is a
+   *   *strict weak ordering*, as defined in the [LessThan Comparable]
+   *   requirements
```
Comment (zasdfgbnm) [inline cub/device/device_merge_sort.cuh:710]: Same here, values don't need to be comparable?

Comment (senior-zero) [reply cub/device/device_merge_sort.cuh]: Thank you for this note! I've fixed everything from #372. The only difference is that I've left the requirement for `ValueIteratorT` being modifiable.

File: cub/device/dispatch/dispatch_merge_sort.cuh [hunk 5]
```diff
@@ -555,21 +559,31 @@ struct DispatchMergeSort : SelectedPolicy
...
+
+      const auto temporary_keys_storage_size =
+        static_cast<std::size_t>(num_items * sizeof(KeyT));
+
+      const auto temporary_values_storage_size =
+        static_cast<std::size_t>(num_items * sizeof(ValueT)) * !KEYS_ONLY;
+
+      const auto virtual_shared_memory_size = vshmem_size(
+        static_cast<std::size_t>(max_shmem),
+        (CUB_NS_QUALIFIER::max)(block_sort_shmem_size, merge_shmem_size),
```
Comment (allisonvacanti) [inline cub/device/dispatch/dispatch_merge_sort.cuh:583]: Just use `cub::` here -- the macro is only needed when used outside of the cub namespace. The other usages of `CUB_NS_MACRO` here can also just be `cub::`.

(The Thrust macro below is correct and should remain)

**Review by allisonvacanti — CHANGES_REQUESTED** [review]

**Pull request merged**